crates/uv/src/commands/tool/upgrade.rs

```diff
@@ -202,6 +218,40 @@ enum UpgradeOutcome {
     NoOp,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum UpgradeConstraint {
+    /// The tool remains pinned to an exact version, so an upgrade was skipped.
+    PinnedVersion { version: Version },
+}
+
+impl UpgradeConstraint {
+    fn print(&self, name: &PackageName, printer: Printer) -> Result<()> {
+        match self {
+            Self::PinnedVersion { version } => {
+                let name_str = name.to_string();
+                let version_str = version.to_string();
+                let reinstall_command = format!("uv tool install {name_str}@latest");
+
+                writeln!(
+                    printer.stderr(),
+                    "hint: `{}` is pinned to `{}` (installed with an exact version pin); reinstall with `{}` to upgrade to a new version.",
+                    name_str.cyan(),
+                    version_str.magenta(),
+                    reinstall_command.green(),
+                )?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct UpgradeReport {
+    outcome: UpgradeOutcome,
+    constraint: Option<UpgradeConstraint>,
+}
```

Comment:
There's still something awkward about this abstraction, but 🤷‍♀️ I'm not having any great ideas so we can leave it until we have another use-case that helps us refactor.

@konstin might have an idea.